File: itest/scripts/VerifyYamlEscapes.java
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
import org.metadatacenter.artifacts.model.core.TextField;
import org.metadatacenter.artifacts.model.tools.YamlSerializer;

/** Companion to verify-yaml-escapes.ts; uses the supplied, built Java artifact library. */
class VerifyYamlEscapes {
  public static void main(String[] args) throws Exception {
    ObjectMapper json = new ObjectMapper();
    ObjectMapper yaml = new ObjectMapper(new YAMLFactory());
    int count = 0;
    for (JsonNode row : json.readTree(System.in)) {
      String value = row.get("value").asText();
      String javaYaml = YamlSerializer.getYAML(TextField.builder()
        .withName("Escape probe").withDescription(value).build(), false, true);
      String scalar = javaYaml.lines().filter(line -> line.startsWith("description:")).findFirst().orElseThrow() + "\n";
      if (!scalar.equals(row.get("tsYaml").asText()))
        throw new AssertionError("Java/TypeScript escape spelling differs for probe " + count);
      if (!value.equals(yaml.readTree(row.get("tsYaml").asText()).get("description").asText()))
        throw new AssertionError("Java changed TypeScript's value for probe " + count);
      JsonNode keyed = yaml.readTree(row.get("tsKeyYaml").asText());
      if (keyed.size() != 1 || !keyed.has(value) || !value.equals(keyed.get(value).asText()))
        throw new AssertionError("Java changed TypeScript's key/value for probe " + count);
      count++;
    }
    System.out.println("Java escape spelling and Java key/value preservation: " + count + " probes passed");
  }
}

File: itest/scripts/verify-yaml-escapes.ts
/**
 * Run with a Java 17 runtime and the built Java artifact library's classpath:
 * npx ts-node itest/scripts/verify-yaml-escapes.ts /path/to/classpath.txt
 * Jest separately checks these Java-generated spellings and TypeScript preservation.
 */
import { readFileSync } from 'node:fs';
import { join } from 'node:path';
import { execFileSync } from 'node:child_process';
import YAML from 'yaml';
import { SimpleYamlSerializer } from '../../src/org/metadatacenter/io/writer/yaml/SimpleYamlSerializer';
import { JsonNode } from '../../src/org/metadatacenter/model/cedar/types/basic-types/JsonNode';

if (!process.argv[2]) throw new Error('Supply the Java artifact library classpath file');
const fixtures: { value: string; javaYaml: string }[] = JSON.parse(
  readFileSync(join(__dirname, '../../test/org/metadatacenter/io/writer/yaml/fixtures/java-escaped-scalars.json'), 'utf8'),
);
const rows = fixtures.map(({ value, javaYaml }) => {
  const tsYaml = SimpleYamlSerializer.serialize({ description: value } as JsonNode);
  const tsKeyYaml = SimpleYamlSerializer.serialize({ [value]: value } as JsonNode);
  if (YAML.parse(javaYaml).description !== value || YAML.parse(tsYaml).description !== value || YAML.parse(tsKeyYaml)[value] !== value)
    throw new Error('TypeScript changed a probe');
  return { value, tsYaml, tsKeyYaml };
});
process.stdout.write(execFileSync('java', ['-cp', readFileSync(process.argv[2], 'utf8').trim(), join(__dirname, 'VerifyYamlEscapes.java')], {
  input: JSON.stringify(rows), encoding: 'utf8',
}));
